WebCoreDecompressionSession should be using asynchronous decoding

jyavenard opened the pull request:
#### 9d02067f5de4b0402f58d6c83039e206439a8a8c
<pre>
WebCoreDecompressionSession should be using asynchronous decoding
<a href="https://bugs.webkit.org/show_bug.cgi?id=292362">https://bugs.webkit.org/show_bug.cgi?id=292362</a>
<a href="https://rdar.apple.com/148016051">rdar://148016051</a>

Reviewed by Youenn Fablet.

Make WebCoreDecompressionSession use a VTDecompressionSession in asynchronous mode.
Also, allows it to handle one to many CMSampleBuffer -&gt; decompressed frame.
while this isn&apos;t a usual scenario, it is possible for the SourceBufferParserWebM
to pack multiple CMSampleBuffer into one, and currently it was only ever returning
the first frame.

No change in observable behaviour, covered by existing tests.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::decodeNextSampleIfNeeded):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSample):
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):
(WebCore::WebCoreDecompressionSession::handleDecompressionOutput):
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):

Canonical link: <a href="https://commits.webkit.org/294421@main">https://commits.webkit.org/294421@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b95358deee08aed5a9675bf5e0465811364eaae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101607 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21275 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11589 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106765 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52241 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103647 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21583 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29772 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77379 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34408 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16670 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91759 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57716 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16496 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9776 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51588 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109118 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21155 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86352 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29101 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87960 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85915 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21893 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30667 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8377 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22892 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28668 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33957 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28479 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31802 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30038 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -225,10 +223,12 @@
                     producer.reject(kVTVideoDecoderBadDataErr);
                 else
-                    producer.resolve(std::exchange(protectedThis->m_lastDecodedSample, { }));
+                    producer.resolve(std::exchange(protectedThis->m_lastDecodedSamples, { }));
```
Is there a case where result is false and m_lastDecodedSamples is not empty?
We could do the std::exchange unconditionally.

jyavenard replied:
> Is there a case where result is false and m_lastDecodedSamples is not empty? We could do the std::exchange unconditionally.

But then you can't return the error. 

m_lastDecodedSamples (and errors) will be cleared at the next call.

99.9% of the time, there's only ever one sample at play. The only exceptions possible today is vp8 with alternative frames, where the first frame returned would actually be null because it's not to be displayed

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -245,25 +245,50 @@
+    auto handler = makeBlockPtr([weakThis = ThreadSafeWeakPtr { *this }, displaying, producer = WTFMove(producer), settled = false, numberOfTimesCalled = 0u, numberOfSamples](OSStatus status, VTDecodeInfoFlags infoFlags, CVImageBufferRef imageBuffer, CMTime presentationTimeStamp, CMTime presentationDuration) mutable {
+        if (settled)
             return;
         if (RefPtr protectedThis = weakThis.get()) {
```
To ease readability, I would tend to do:
```RefPtr protectedThis = weakThis.get();
if (!protectedThis) {
    settled = true;
    producer.reject(0);
    return;
}```

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -245,25 +245,50 @@
+                        protectedThis->m_lastDecodedSamples.append(WTFMove(*result));
+                        producer.resolve(std::exchange(protectedThis->m_lastDecodedSamples, { }));
+                    }
+                });
```
We could `return;` here to not have to do if the `else` below.

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -245,25 +245,50 @@
+                    }
+                    auto result = protectedThis->handleDecompressionOutput(displaying, status, infoFlags, imageBuffer.get(), presentationTimeStamp, presentationDuration);
+                    if (!result)
+                        producer.reject(result.error());
```
Here we could set back `m_lastDecodedSamples` to { }?

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -296,24 +321,22 @@
 
     if (isNonRecoverableError(status)) {
         RELEASE_LOG_ERROR(Media, "Video sample decompression failed with error:%d", int(status));
-        m_lastDecodingError = status;
```
Does `handleDecompressionOutput` needs to be a method, can it be a free function?

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -356,13 +377,19 @@
-                protectedThis->handleDecompressionOutput(protectedThis->m_pendingDecodeData->displaying, noErr, 0, result->frame->pixelBuffer(), presentationTime, presentationDuration);
+                auto sampleResult = protectedThis->handleDecompressionOutput(protectedThis->m_pendingDecodeData->displaying, noErr, 0, result->frame->pixelBuffer(), presentationTime, presentationDuration);
+                if (!sampleResult)
+                    protectedThis->m_lastDecodingError = sampleResult.error();
```
Ditto here maybe.

youennf reviewed (approved):
LGTM

jyavenard committed:
WebCoreDecompressionSession should be using asynchronous decoding
https://bugs.webkit.org/show_bug.cgi?id=292362
rdar://148016051

Reviewed by Youenn Fablet.

Make WebCoreDecompressionSession use a VTDecompressionSession in asynchronous mode.
Also, allows it to handle one to many CMSampleBuffer -> decompressed frame.
while this isn't a usual scenario, it is possible for the SourceBufferParserWebM
to pack multiple CMSampleBuffer into one, and currently it was only ever returning
the first frame.

No change in observable behaviour, covered by existing tests.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::decodeNextSampleIfNeeded):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSample):
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):
(WebCore::WebCoreDecompressionSession::handleDecompressionOutput):
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):

Canonical link: https://commits.webkit.org/294421@main
